index.js

```diff
@@ -6,63 +6,54 @@ var glob = require('glob');
 var arrify = require('arrify');
 var pify = require('pify');
 
-function sortPatterns(patterns) {
-	patterns = arrify(patterns);
+var globP = pify(glob, Promise).bind(glob);
 
-	var positives = [];
-	var negatives = [];
+function isNegative(pattern) {
+	return pattern[0] === '!';
+}
 
-	patterns.forEach(function (pattern, index) {
-		var isNegative = pattern[0] === '!';
-		(isNegative ? negatives : positives).push({
-			index: index,
-			pattern: isNegative ? pattern.slice(1) : pattern
-		});
-	});
+function generateGlobTasks(patterns, opts) {
+	var globTasks = [];
 
-	return {
-		positives: positives,
-		negatives: negatives
-	};
-}
+	patterns = arrify(patterns);
+	opts = objectAssign({ignore: []}, opts);
 
-function setIgnore(opts, negatives, positiveIndex) {
-	opts = objectAssign({}, opts);
+	patterns.forEach(function (pattern, i) {
+		if (isNegative(pattern)) {
+			return;
+		}
 
-	var negativePatterns = negatives.filter(function (negative) {
-		return negative.index > positiveIndex;
-	}).map(function (negative) {
-		return negative.pattern;
+		var ignore = patterns.slice(i).filter(isNegative).map(function (pattern) {
+			return pattern.slice(1);
+		});
+
+		globTasks.push({
+			pattern: pattern,
+			opts: objectAssign({}, opts, {
+				ignore: opts.ignore.concat(ignore)
+			})
+		});
 	});
 
-	opts.ignore = (opts.ignore || []).concat(negativePatterns);
-	return opts;
+	return globTasks;
 }
 
 module.exports = function (patterns, opts) {
-	var sortedPatterns = sortPatterns(patterns);
-	opts = opts || {};
```

Comment:
(I know how `Object.assign`/`objectAssign` works, just making sure your reasoning for removing this is because you're always making a shallow copy of the options object using `objectAssign` before modifying it)